Source/WebCore/accessibility/AXTextMarker.cpp

```diff
@@ -858,6 +941,53 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
     RELEASE_ASSERT_NOT_REACHED();
 }
 
+AXTextMarkerRange AXTextMarker::wordRange(WordRangeType type) const
+{
+    if (!isValid())
+        return { { }, { } };
+
+    AXTextMarker startMarker, endMarker;
+
+    if (type == WordRangeType::Right) {
+        startMarker = nextWordStart();
+        endMarker = startMarker.nextWordEnd();
+        auto resultRange = AXTextMarkerRange { startMarker, endMarker };
+        auto resultRangeString = resultRange.toString();
+
+        // A newline should not be returned as the right word (if we are at the end of a paragraph)
+        if (startMarker == *this && resultRangeString.length() && resultRangeString[0] == '\n')
+            return { WTFMove(*this), WTFMove(*this) };
+    } else {
+        startMarker = previousWordStart();
+        if (startMarker == *this)
+            return { startMarker, startMarker };
+
+        endMarker = startMarker.nextWordEnd();
+        auto resultRange = AXTextMarkerRange { startMarker, endMarker };
+        auto resultRangeString = resultRange.toString();
+
+        // Check forward from the current offset for the next word start.
+        // If we are on the start of a word (next marker is a word, previous is whitespace), left word should be nil.
+        auto oneAheadStartMarker = nextWordStart();
+        auto nextMarker = findMarker(AXDirection::Next, false, true);
+
+        // There are two cases that will happen to indicate we are on the left word boundary:
+        // (1) The word break iterator returns the whitespace to the left of the word, and the end marker is the equivalent to the next starting word (handled by `isWhitespceAndEndsOnLeftWordBoundary`.
+        // (2) Thre word break iterator returns a word end before the current marker, and the next word starts at the visual equivalent to the current marker.
+        bool isWhitespaceAndEndsOnLeftWordBoundary = equivalentTextPosition(endMarker) && isASCIIWhitespace(resultRangeString[resultRangeString.length() - 1]);
```

Comment:
`isASCIIWhitespace(resultRangeString[resultRangeString.length() - 1]` — do we want to guard against zero-length `resultRangeString` strings?

Reply:
ah yeah, I'll add a check here

Reply:
Just noticed a typo in this bit too: "Thre"